include/envoy/config/typed_config.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * Base class for config defined by a typed proto message.
+ */
+class TypedConfig {
+public:
+  /**
+   * @return ProtobufTypes::MessagePtr create empty config proto message for v2. The config, which
```

Comment:
Given that folks are going to be starting to move to v3 in early 2020, do we want to do a lot of work to make this supported in v2?

Reply:
We need some bridge to associate a protobuf type with a factory. We probably could invert this, and just use `type` to instantiate a proto. Either way works, I'm aiming for minimal disruption in this PR.

Reply:
What do you think about getting https://github.com/envoyproxy/envoy/issues/8933 landed? In this world, all extensions are typed..

Reply:
Yes, that would be wonderful. I was going to special case empty config but would be better without that.